lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/BuildJsonHttpDeserializer.kt

```diff
@@ -57,36 +41,60 @@ internal fun <T> buildJsonHttpDeserializer(
     deserializationJson: Json,
 ): HttpResponseDeserializer<T> {
     return Function { response ->
-        if (!response.isSuccessful) {
-            throw RequestFailedException(
-                payload = response.body,
-            )
-        }
+        validateResponseSuccess(response)
+
         val jsonRpcResponse =
             deserializationJson.decodeFromString(
                 JsonRpcResponse.serializer(deserializationStrategy),
                 response.body,
             )
 
-        extractResult(jsonRpcResponse, response.body)
+        if (jsonRpcResponse.error != null) {
+            throw RpcRequestFailedException(
+                code = jsonRpcResponse.error.code,
+                message = jsonRpcResponse.error.message,
+                data = jsonRpcResponse.error.data,
+                payload = response.body,
+            )
+        }
+
+        if (jsonRpcResponse.result == null) {
+            throw RequestFailedException(message = "Response did not contain a result", payload = response.body)
+        }
+        jsonRpcResponse.result
     }
 }
 
-internal fun <T> buildJsonBatchHttpDeserializer(
+internal fun <T>buildJsonHttpBatchDeserializer(
     deserializationStrategies: List<KSerializer<T>>,
     deserializationJson: Json,
-): HttpResponseDeserializer<List<T>> {
-    // TODO: In case of batch request, exception should not be thrown.
-    // Instead, we want to return wrapped responses.
-    // This enables access to successful results or encountered errors.
-
+): HttpResponseDeserializer<List<Result<T>>> {
     return Function { response ->
-        if (!response.isSuccessful) {
-            throw RequestFailedException(
-                payload = response.body,
+        validateResponseSuccess(response)
+
+        val jsonResponses = Json.parseToJsonElement(response.body).jsonArray
+        val responses = jsonResponses.map {
+            deserializationJson.decodeFromJsonElement(
+                JsonRpcResponse.serializer(deserializationStrategies[it.jsonObject["id"]!!.jsonPrimitive.int]),
+                it,
             )
         }
 
+        val results = responses.sortedBy { it.id }.zip(jsonResponses)
```

Comment:
is there a possible performance hit with that?

Reply:
I can surely prepare some performance tests on that. Wdyt?

Reply:
Update: sorting has been completely removed as it was unnecessary. Correct assignment of deserializers to json responses is done by accessing element with adequate index from deserializers list - index is the id coming from json. Full implementation [here](https://github.com/software-mansion/starknet-jvm/blob/a8cb6a320165a1c53e389f90e30615e618f00623/lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/BuildJsonHttpDeserializer.kt#L81).

Reply:
Update: I've improved sorting the results. Using a `MutableList` and directly placing items based on their ids will has better time complexity `O(n)` than using `sortedBy` which has `O(nlogn)`.
Implementation can be found [here](https://github.com/software-mansion/starknet-jvm/blob/ce6d87a1366371591ec02809287645833ed0e491/lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/BuildJsonHttpDeserializer.kt#L104)
 Wdyt?

Reply:
Update: resolving this as the implementation ha been approved in upstream branch.